### DISCUSSION TITLE
Use topology for routes

### DISCUSSION THREAD

**Discussion by dumpfheimer** [zigpy/zigpy#1257] [Ideas]
Hi!

I started playing around with finding good routes in the application instead of the coordinator a few months ago.
I ran out of time for further development, but I think I have a good compromise that could take a lot of load off of the coordinator.

On the ZNP type coordinator I often found that on "large" networks (I have > 100 Devices) the coordinator seems to have insufficient memory to store routes to all end devices. This caused a very noticable latency when a light was to be reached, that had fallen out of that mapping. Once solution to this problem could be to tell the coordinator up front which route to take, skipping the route discovery on the coordinator all together.

I tried to implement this, and did succeed, but the performance for longer routes was horrible (it sometimes took a couple hundred ms to calculate/find the best route for longer routes).

But, finding 0-hop and 1-hop routes from the coordinator to devices is very fast and fairly simple. And every device, that is known to be within the range of the coordinator and 1 hop away will be one device less that needs to be in the memory mapping of the coordinator, thus freeing memory up for those, that need multiple hops.

Well, at least this is how I believe it works. If somebody with more technical knowledge knows better, please jump in and correct me.

**TL;DR**

I developed a route discovery based on the topology of the zigbee network, that can find 0-hop and 1-hop routes.

Source code:
[1st commit](https://github.com/dumpfheimer/zigpy/commit/c39cddac2bd215b15b8c26297197798b37dca73b)
[Repository](https://github.com/dumpfheimer/zigpy/tree/dumpfheimer/enhanced_source_routing)

To enable it, edit your ha configuration.yaml
```
zha:
    zigpy_config:
        enhanced_source_routing: True
```

This can be combined with the default source_routing funcionality, which could step in if the topology route does not find a route. But source routing seems to tend to be unstable during "warm up" of the network.

Hope this sparks some interest or ideas or might even fix some problems for someone.

Any feedback is appreciated.

**Comment by MattWestb** [comment]
One PM about TI coordinators is that the have tuning the broadcast table out of range for getting fast and nice broadcast but the underlying 802.15.4 network is blocking it for broadcast storm protection.
Then its happening the network cant doing router discovery OK then broadcast commands is being blocked and most or router discovery is made with broadcast.
Also with the 2023 fix of disabled Zigbee 3 end device handing in the coordinator routing part they was getting large problems with addressing and memory handling then most devices was joined as Zigbee HA 1.X before and not Zigbee 3 but i think all memory problems is the same in the ground.

**Reply by dumpfheimer** [reply]
Hi Matt,
thanks for checking in.

What was "the 2023 fix"?

Do I understand you correctly, that route discoveries mean more broadcast messages, which means blocked air time? That means finding the optimal route in application would have a positive effect on network performance, right?

**Reply by MattWestb** [reply]
The first point in the change log 
https://github.com/Koenkk/Z-Stack-firmware/blob/master/coordinator/Z-Stack_3.x.0/CHANGELOG.md

= **they was never handling Zigbee 3 end device** like IKEA controllers was not able to have the coordinator as parent and other device that can doing Zigbee HA 1,.2 paring was OK = not using updated link key and so on.

The `MAX_BCAST 30` is braking the router discovery if doing little broadcasts and my issue https://github.com/Koenkk/Z-Stack-firmware/issues/443 is closed as not wanted.

Read the issue and understanding **if** the IEEE 892.15.4 is blocking for broadcast storm also the router discovers is not working and its not one Zigbee problem its how underlying 802.15.4 is working and cant being caged without braking it completely.

Do one sniff one one network that  is having some router with our without source routing and you see how much broadcast is being made but most is only one hope but its not helping if one router is blocked its ignoring the frame = lost.

**Comment by Hedda** [comment]
@dumpfheimer I believe that the deCONZ (Phoscon) application support using a similar enhanced source routing concept, including both automatic and manual manual source routing, check out this related discussion-> https://github.com/zigpy/zigpy-deconz/issues/180

**Reply by dumpfheimer** [reply]
I just quickly read over it, but it seems like they are using quite normal source routing, am I mistaken?

**Reply by dumpfheimer** [reply]
Thanks for the link!

**Reply by MattWestb** [reply]
Im very sorry but DEs **manual source routing** is braking the self healing functionality of the network and is nothing that shall being used if you like have one good working network.
It can being used routing traffic around one bad router but its better taking the bad device out of the network (like OSRAM plugs).
I don t knowing if the manual in implanted but it was on the list before then was blocking my as one payed customer that like have its support of there product so i dont helping them after that then they is not paying for it.
